Drop CPAN modules unused in any currently supported distro from RPM

mavit opened the pull request:
This is based on the following assumptions:

- We only care about Fedora, RHEL (and rebuilds), openSUSE Leap, and SLES.
- We only care about distribution versions that are still supported by their vendor.

If these assumptions are incorrect, then this pull request should, of course, be rejected.  I imagine there are some insights in the analytics data, but I don’t think they are publicly available?

mavit committed:
Drop CPAN modules unused in any currently supported distro from RPM

This is based on the following assumptions:

- We only care about Fedora, RHEL (and rebuilds), openSUSE Leap, and SLES.
- We only care about distribution versions that are still supported by their vendor.

Signed-off-by: Peter Oliver <git@mavit.org.uk>

mw9 commented:
This does appear to have an impact in `buildDebian`. Is that intentional ?

I haven’t worked through the significance, though.

michaelherger commented:
This is RPM only, right?

Would you have a matrix of OS/version/Perl version? I could try to get numbers from the data we have.

michaelherger commented:
Perl versions/platform for RPM based systems seen in the last three months:

| Count | Perl Version | Architecture |
|------:|-------------|--------------|
| 123 | 5.36.1 | x86_64-linux |
| 54 | 5.40.3 | x86_64-linux |
| 53 | 5.32.1 | x86_64-linux |
| 50 | 5.36.0 | x86_64-linux |
| 49 | 5.34.1 | armv7l-linux |
| 33 | 5.28.2 | x86_64-linux |
| 25 | 5.26.3 | x86_64-linux |
| 17 | 5.16.3 | x86_64-linux |
| 14 | 5.34.1 | x86_64-linux |
| 10 | 5.40.2 | x86_64-linux |
| 9 | 5.38.4 | x86_64-linux |
| 8 | 5.24.3 | x86_64-linux |
| 8 | 5.38.2 | x86_64-linux |
| 3 | 5.22.2 | x86_64-linux |
| 2 | 5.36.3 | x86_64-linux |
| 2 | 5.40.0 | x86_64-linux |
| 2 | 5.40.1 | x86_64-linux |
| 2 | 5.40.3 | aarch64-linux |
| 1 | 5.16.3 | i686-linux |
| 1 | 5.26.2 | x86_64-linux |
| 1 | 5.30.3 | x86_64-linux |
| 1 | 5.34.1 | aarch64-linux |

It's relatively small numbers... and hardly any aarch64!

mavit commented:
> This does appear to have an impact in `buildDebian`. Is that intentional ?

It's not intentional.  My reading of the code is that it shouldn't affect Debian packages in practice, because we only build them on Arm and Intel platforms, which each have their own set of exclusions.

mavit commented:
Huh, there was me hoping we wouldn't need Perl 5.36 any more, but there it is, right at the top of the popularity list.  I wonder why that is.  Could be Fedora 37, maybe, or something based on it.

In any case, I will redo this patch based on that list...

mavit committed:
Tweak RPM CPAN exclusions based on usage stats

https://github.com/LMS-Community/slimserver-platforms/pull/99?#issuecomment-3423325041

michaelherger reviewed (approved):
Thanks! I think I'm going to review what we're bundling with the .deb packages, too.

michaelherger commented:
Oh, btw: could you please look into the DCO sign-off stuff? Thanks!

mw9 commented:
> Thanks! I think I'm going to review what we're bundling with the .deb packages, too.

At present the Debian build does not **require** that an architecture be provided. Perhaps that might be modified to require that **one** architecture be provided.

These [lines](https://github.com/LMS-Community/slimserver-platforms/blob/c8deb34f193b12a087ef1ccf64a72e89ac2862f4/buildme.pl#L682C1-L685C1) in `buildDebian` might then be modified from:
```
	else {
		removeExclusions($dirsToExcludeForLinuxPackage);
	}
```
to:
```
	else {
		print "INFO:  A Debian build requires an architecture \n";
		exit 1;
	}
```
Or something like that.

Just a thought.

michaelherger commented:
@mw9 we actually still build a `.deb` with no architecture. It's the `..._all.deb` version. I actually dropped the `i386` build earlier today, as well as Perl 5.20-5.26 from Debian packages - except for the `all` build. That way those few still dealing with some outdated platform can still use a package, though it'll be a bigger download.

mw9 commented:
> @mw9 we actually still build a `.deb` with no architecture. It's the `..._all.deb` version.

I had missed that.

I had also missed that `$dirsToExcludeForLinuxPackage` now remains unchanged in the final form of the PR. So there is no longer any impact to `buildDebian`. Which is what I had first noticed.

So my comment was simply noise... apologies.

mavit commented:
> I actually dropped the `i386` build earlier today

Does Tuesday's i686 `.deb` need deleting from the downloads website, then?

michaelherger commented:
> > I actually dropped the `i386` build earlier today
> 
> Does Tuesday's i686 `.deb` need deleting from the downloads website, then?

It should disappear eventually. We keep three builds around. So once we've had three builds after the removal it should be gone from the web site. But I'll double check whether anything else needs to be done. Thanks for the heads up!
